File: tools/town_generator/reference/mfcg-clean/geometry/DCEL.js
/**
 * DCEL.js - Doubly Connected Edge List
 *
 * A topological data structure for representing planar subdivisions.
 * Used for Voronoi diagrams and city patch topology.
 */

import { Point } from '../core/Point.js';

/**
 * Vertex in the DCEL
 */
export class Vertex {
    /**
     * @param {Point} point - Position of vertex
     */
    constructor(point) {
        this.point = point;
        /** @type {HalfEdge[]} */
        this.edges = []; // All half-edges originating from this vertex
    }

    get x() { return this.point.x; }
    get y() { return this.point.y; }
}

/**
 * Half-edge in the DCEL
 */
export class HalfEdge {
    constructor() {
        /** @type {Vertex} */
        this.origin = null;
        /** @type {HalfEdge} */
        this.twin = null;
        /** @type {HalfEdge} */
        this.next = null;
        /** @type {HalfEdge} */
        this.prev = null;
        /** @type {Face} */
        this.face = null;
        /** @type {*} */
        this.data = null; // Custom data (e.g., edge type)
    }

    /**
     * Get destination vertex
     * @returns {Vertex}
     */
    get destination() {
        return this.next ? this.next.origin : null;
    }

    /**
     * Get length of edge
     * @returns {number}
     */
    get length() {
        if (!this.origin || !this.destination) return 0;
        return Point.distance(this.origin.point, this.destination.point);
    }
}

/**
 * Face in the DCEL
 */
export class Face {
    constructor() {
        /** @type {HalfEdge} */
        this.halfEdge = null; // One half-edge on the boundary
        /** @type {*} */
        this.data = null; // Custom data (e.g., Cell)
    }

    /**
     * Get polygon for this face
     * @returns {Point[]}
     */
    getPoly() {
        const poly = [];
        let edge = this.halfEdge;
        if (!edge) return poly;

        do {
            poly.push(edge.origin.point);
            edge = edge.next;
        } while (edge && edge !== this.halfEdge);

        return poly;
    }

    /**
     * Iterate over edges of this face
     * @yields {HalfEdge}
     */
    *edges() {
        let edge = this.halfEdge;
        if (!edge) return;

        do {
            yield edge;
            edge = edge.next;
        } while (edge && edge !== this.halfEdge);
    }

    /**
     * Get all vertices of this face
     * @returns {Vertex[]}
     */
    vertices() {
        const verts = [];
        for (const edge of this.edges()) {
            verts.push(edge.origin);
        }
        return verts;
    }
}

/**
 * Doubly Connected Edge List
 */
export class DCEL {
    /**
     * Create DCEL from polygons
     * @param {Point[][]} polygons - Array of polygon point arrays
     */
    constructor(polygons = []) {
        /** @type {Map<Point, Vertex>} */
        this.vertices = new Map();
        /** @type {HalfEdge[]} */
        this.edges = [];
        /** @type {Face[]} */
        this.faces = [];

        if (polygons.length > 0) {
            this.buildFromPolygons(polygons);
        }
    }

    /**
     * Build DCEL from array of polygons
     * @param {Point[][]} polygons
     */
    buildFromPolygons(polygons) {
        // Create vertices
        for (const poly of polygons) {
            for (const point of poly) {
                if (!this.vertices.has(point)) {
                    this.vertices.set(point, new Vertex(point));
                }
            }
        }

        // Map for finding twin edges
        const edgeMap = new Map();

        const edgeKey = (p1, p2) => {
            const k1 = `${p1.x.toFixed(6)},${p1.y.toFixed(6)}`;
            const k2 = `${p2.x.toFixed(6)},${p2.y.toFixed(6)}`;
            return `${k1}->${k2}`;
        };

        // Create faces and edges
        for (const poly of polygons) {
            const face = new Face();
            this.faces.push(face);

            const faceEdges = [];
            const n = poly.length;

            for (let i = 0; i < n; i++) {
                const p1 = poly[i];
                const p2 = poly[(i + 1) % n];

                const edge = new HalfEdge();
                edge.origin = this.vertices.get(p1);
                edge.origin.edges.push(edge);
                edge.face = face;

                faceEdges.push(edge);
                this.edges.push(edge);

                // Store for twin lookup
                edgeMap.set(edgeKey(p1, p2), edge);
            }

            // Link edges in cycle
            for (let i = 0; i < faceEdges.length; i++) {
                faceEdges[i].next = faceEdges[(i + 1) % faceEdges.length];
                faceEdges[i].prev = faceEdges[(i + faceEdges.length - 1) % faceEdges.length];
            }

            face.halfEdge = faceEdges[0];
            face.data = { shape: poly };
        }

        // Link twins
        for (const poly of polygons) {
            const n = poly.length;
            for (let i = 0; i < n; i++) {
                const p1 = poly[i];
                const p2 = poly[(i + 1) % n];

                const edge = edgeMap.get(edgeKey(p1, p2));
                const twin = edgeMap.get(edgeKey(p2, p1));

                if (edge && twin) {
                    edge.twin = twin;
                    twin.twin = edge;
                }
            }
        }
    }

    /**
     * Get vertex for point (or create if not exists)
     * @param {Point} point
     * @returns {Vertex}
     */
    getVertex(point) {
        if (!this.vertices.has(point)) {
            this.vertices.set(point, new Vertex(point));
        }
        return this.vertices.get(point);
    }

    /**
     * Find circumference edges of a set of faces
     * @param {HalfEdge|null} startEdge - Optional start edge
     * @param {Face[]} faces - Faces to find circumference of
     * @returns {HalfEdge[]}
     */
    static circumference(startEdge, faces) {
        const faceSet = new Set(faces);
        const boundaryEdges = [];

        for (const face of faces) {
            for (const edge of face.edges()) {
                // Edge is on boundary if twin is null or twin's face is not in set
                if (!edge.twin || !faceSet.has(edge.twin.face)) {
                    boundaryEdges.push(edge);
                }
            }
        }

        if (boundaryEdges.length === 0) return [];

        // Sort edges into a cycle
        const result = [];
        let current = startEdge && boundaryEdges.includes(startEdge)
            ? startEdge
            : boundaryEdges[0];

        const visited = new Set();
        while (current && !visited.has(current)) {
            visited.add(current);
            result.push(current);

            // Find next boundary edge
            let next = current.next;
            while (next && !boundaryEdges.includes(next)) {
                if (next.twin) {
                    next = next.twin.next;
                } else {
                    break;
                }
            }
            current = boundaryEdges.includes(next) ? next : null;
        }

        return result;
    }

    /**
     * Convert vertex array to edge chain
     * @param {Vertex[]} vertices
     * @returns {HalfEdge[]}
     */
    vertices2chain(vertices) {
        const chain = [];
        for (let i = 0; i < vertices.length - 1; i++) {
            const v1 = vertices[i];
            const v2 = vertices[i + 1];

            // Find edge from v1 to v2
            for (const edge of v1.edges) {
                if (edge.destination === v2) {
                    chain.push(edge);
                    break;
                }
            }
        }
        return chain;
    }

    /**
     * Split faces into connected components
     * @param {Face[]} faces
     * @returns {Face[][]}
     */
    static split(faces) {
        const faceSet = new Set(faces);
        const visited = new Set();
        const components = [];

        for (const face of faces) {
            if (visited.has(face)) continue;

            const component = [];
            const queue = [face];

            while (queue.length > 0) {
                const current = queue.shift();
                if (visited.has(current)) continue;
                visited.add(current);
                component.push(current);

                // Add adjacent faces in the set
                for (const edge of current.edges()) {
                    if (edge.twin && faceSet.has(edge.twin.face) && !visited.has(edge.twin.face)) {
                        queue.push(edge.twin.face);
                    }
                }
            }

            components.push(component);
        }

        return components;
    }

    /**
     * Collapse an edge, merging its endpoints
     * @param {HalfEdge} edge
     * @returns {{vertex: Vertex, edges: HalfEdge[]}}
     */
    collapseEdge(edge) {
        const v1 = edge.origin;
        const v2 = edge.destination;
        const midpoint = Point.midpoint(v1.point, v2.point);

        // Update v1's position to midpoint
        v1.point.x = midpoint.x;
        v1.point.y = midpoint.y;

        // Move all v2's edges to v1
        for (const e of v2.edges) {
            e.origin = v1;
            v1.edges.push(e);
        }
        v2.edges = [];

        // Remove the collapsed edge from the cycle
        if (edge.prev) edge.prev.next = edge.next;
        if (edge.next) edge.next.prev = edge.prev;

        if (edge.twin) {
            if (edge.twin.prev) edge.twin.prev.next = edge.twin.next;
            if (edge.twin.next) edge.twin.next.prev = edge.twin.prev;
        }

        // Update face pointers if needed
        if (edge.face.halfEdge === edge) {
            edge.face.halfEdge = edge.next !== edge ? edge.next : null;
        }
        if (edge.twin && edge.twin.face.halfEdge === edge.twin) {
            edge.twin.face.halfEdge = edge.twin.next !== edge.twin ? edge.twin.next : null;
        }

        // Remove from edges list
        const idx = this.edges.indexOf(edge);
        if (idx >= 0) this.edges.splice(idx, 1);
        if (edge.twin) {
            const twinIdx = this.edges.indexOf(edge.twin);
            if (twinIdx >= 0) this.edges.splice(twinIdx, 1);
        }

        // Remove v2 from vertices
        this.vertices.delete(v2.point);

        // Return affected edges for face update
        const affectedEdges = [];
        if (edge.face) affectedEdges.push(edge.next || edge.prev);
        if (edge.twin && edge.twin.face) affectedEdges.push(edge.twin.next || edge.twin.prev);

        return { vertex: v1, edges: affectedEdges };
    }

    /**
     * Split an edge at its midpoint
     * @param {HalfEdge} edge
     * @returns {Vertex}
     */
    splitEdge(edge) {
        const midpoint = Point.midpoint(edge.origin.point, edge.destination.point);
        const newVertex = new Vertex(midpoint);
        this.vertices.set(midpoint, newVertex);

        const newEdge = new HalfEdge();
        newEdge.origin = newVertex;
        newEdge.face = edge.face;
        newEdge.next = edge.next;
        newEdge.prev = edge;
        if (edge.next) edge.next.prev = newEdge;
        edge.next = newEdge;

        newVertex.edges.push(newEdge);
        this.edges.push(newEdge);

        if (edge.twin) {
            const newTwin = new HalfEdge();
            newTwin.origin = newVertex;
            newTwin.face = edge.twin.face;
            newTwin.twin = edge;
            edge.twin.twin = newEdge;
            newEdge.twin = edge.twin;
            edge.twin = newTwin;

            newTwin.next = edge.twin.next;
            newTwin.prev = edge.twin;
            if (edge.twin.next) edge.twin.next.prev = newTwin;
            edge.twin.next = newTwin;

            newVertex.edges.push(newTwin);
            this.edges.push(newTwin);
        }

        return newVertex;
    }
}

/**
 * Edge chain utilities
 */
export class EdgeChain {
    /**
     * Convert edge chain to polygon
     * @param {HalfEdge[]} chain
     * @returns {Point[]}
     */
    static toPoly(chain) {
        return chain.map(edge => edge.origin.point);
    }

    /**
     * Convert edge chain to polyline (includes last point)
     * @param {HalfEdge[]} chain
     * @returns {Point[]}
     */
    static toPolyline(chain) {
        if (chain.length === 0) return [];
        const points = chain.map(edge => edge.origin.point);
        if (chain[chain.length - 1].destination) {
            points.push(chain[chain.length - 1].destination.point);
        }
        return points;
    }

    /**
     * Get all vertices in edge chain
     * @param {HalfEdge[]} chain
     * @returns {Vertex[]}
     */
    static vertices(chain) {
        return chain.map(edge => edge.origin);
    }

    /**
     * Assign data to all edges in chain
     * @param {HalfEdge[]} chain
     * @param {*} data
     * @param {boolean} [overwrite=true]
     */
    static assignData(chain, data, overwrite = true) {
        for (const edge of chain) {
            if (overwrite || edge.data === null) {
                edge.data = data;
            }
        }
    }

    /**
     * Find edge by origin vertex
     * @param {HalfEdge[]} chain
     * @param {Vertex} vertex
     * @returns {HalfEdge|null}
     */
    static edgeByOrigin(chain, vertex) {
        for (const edge of chain) {
            if (edge.origin === vertex) return edge;
        }
        return null;
    }
}

File: tools/town_generator/src/geom/DCEL.cpp
#include "town_generator/geom/DCEL.h"
#include "town_generator/geom/EdgeChain.h"
#include <algorithm>
#include <queue>
#include <cmath>

namespace town_generator {
namespace geom {

// =============================================================================
// HalfEdge Implementation
// =============================================================================

double HalfEdge::length() const {
    auto dest = destination();
    if (!origin || !dest) return 0.0;
    return Point::distance(*origin->point, *dest->point);
}

// =============================================================================
// Face Implementation
// =============================================================================

std::vector<Point> Face::getPoly() const {
    std::vector<Point> poly;
    if (!halfEdge) return poly;

    auto edge = halfEdge;
    do {
        poly.push_back(*edge->origin->point);
        edge = edge->next;
    } while (edge && edge != halfEdge);

    return poly;
}

std::vector<PointPtr> Face::getPolyPtrs() const {
    std::vector<PointPtr> poly;
    if (!halfEdge) return poly;

    auto edge = halfEdge;
    do {
        poly.push_back(edge->origin->point);
        edge = edge->next;
    } while (edge && edge != halfEdge);

    return poly;
}

std::vector<VertexPtr> Face::vertices() const {
    std::vector<VertexPtr> verts;
    if (!halfEdge) return verts;

    auto edge = halfEdge;
    do {
        verts.push_back(edge->origin);
        edge = edge->next;
    } while (edge && edge != halfEdge);

    return verts;
}

size_t Face::edgeCount() const {
    if (!halfEdge) return 0;

    size_t count = 0;
    auto edge = halfEdge;
    do {
        ++count;
        edge = edge->next;
    } while (edge && edge != halfEdge);

    return count;
}

double Face::area() const {
    // Shoelace formula
    auto poly = getPoly();
    if (poly.size() < 3) return 0.0;

    double sum = 0.0;
    size_t n = poly.size();
    for (size_t i = 0; i < n; ++i) {
        const Point& p1 = poly[i];
        const Point& p2 = poly[(i + 1) % n];
        sum += (p1.x * p2.y - p2.x * p1.y);
    }
    return std::abs(sum) / 2.0;
}

Point Face::centroid() const {
    auto poly = getPoly();
    if (poly.empty()) return Point(0, 0);

    double cx = 0, cy = 0;
    for (const auto& p : poly) {
        cx += p.x;
        cy += p.y;
    }
    return Point(cx / poly.size(), cy / poly.size());
}

// =============================================================================
// DCEL Implementation
// =============================================================================

DCEL::DCEL(const std::vector<Polygon>& polygons) {
    buildFromPolygons(polygons);
}

void DCEL::buildFromPolygons(const std::vector<Polygon>& polygons) {
    // Clear existing data
    vertices.clear();
    edges.clear();
    faces.clear();

    // Create vertices from polygon points (uses existing PointPtrs)
    for (const auto& poly : polygons) {
        for (size_t i = 0; i < poly.length(); ++i) {
            PointPtr p = poly.ptr(i);
            getOrCreateVertex(p);
        }
    }

    // Edge key for twin lookup: maps (p1, p2) -> HalfEdge*
    // Using raw pointers for map key since PointPtr comparison is by identity
    std::map<std::pair<Point*, Point*>, HalfEdgePtr> edgeMap;

    // Create faces and edges
    for (const auto& poly : polygons) {
        if (poly.length() < 3) continue;

        auto face = std::make_shared<Face>();
        std::vector<HalfEdgePtr> faceEdges;

        for (size_t i = 0; i < poly.length(); ++i) {
            PointPtr p1 = poly.ptr(i);
            PointPtr p2 = poly.ptr((i + 1) % poly.length());

            auto edge = std::make_shared<HalfEdge>();
            edge->origin = vertices[p1];
            edge->origin->addEdge(edge);
            edge->face = face;

            faceEdges.push_back(edge);
            edgeMap[{p1.get(), p2.get()}] = edge;
            edges.push_back(edge);
        }

        // Link edges in cycle (next/prev)
        // Note: next is shared_ptr (ownership chain), prev is weak_ptr (back-ref)
        for (size_t i = 0; i < faceEdges.size(); ++i) {
            faceEdges[i]->next = faceEdges[(i + 1) % faceEdges.size()];
            faceEdges[i]->prev = faceEdges[(i + faceEdges.size() - 1) % faceEdges.size()];
        }

        face->halfEdge = faceEdges[0];
        faces.push_back(face);
    }

    // Link twin edges (edges with reversed direction)
    for (const auto& [key, edge] : edgeMap) {
        auto twinKey = std::make_pair(key.second, key.first);
        auto it = edgeMap.find(twinKey);
        if (it != edgeMap.end()) {
            edge->twin = it->second;
        }
    }
}

VertexPtr DCEL::getVertex(const PointPtr& point) const {
    auto it = vertices.find(point);
    return (it != vertices.end()) ? it->second : nullptr;
}

VertexPtr DCEL::getOrCreateVertex(PointPtr point) {
    auto it = vertices.find(point);
    if (it != vertices.end()) {
        return it->second;
    }

    auto vertex = std::make_shared<Vertex>(point);
    vertices[point] = vertex;
    return vertex;
}

void DCEL::cleanup() {
    for (auto& [point, vertex] : vertices) {
        vertex->cleanupEdges();
    }
}

// =============================================================================
// Topology Operations
// =============================================================================

std::vector<HalfEdgePtr> DCEL::circumference(const HalfEdgePtr& startEdge,
                                              const std::vector<FacePtr>& faceList) {
    if (faceList.empty()) return {};

    std::set<Face*> faceSet;
    for (const auto& f : faceList) {
        if (f) faceSet.insert(f.get());
    }

    std::vector<HalfEdgePtr> boundaryEdges;

    // Find all edges on the boundary
    for (const auto& face : faceList) {
        if (!face || !face->halfEdge) continue;

        auto edge = face->halfEdge;
        do {
            // Edge is on boundary if twin is null or twin's face is not in set
            auto twinEdge = edge->getTwin();
            if (!twinEdge) {
                boundaryEdges.push_back(edge);
            } else {
                auto twinFace = twinEdge->getFace();
                if (!twinFace || faceSet.find(twinFace.get()) == faceSet.end()) {
                    boundaryEdges.push_back(edge);
                }
            }
            edge = edge->next;
        } while (edge && edge != face->halfEdge);
    }

    if (boundaryEdges.empty()) return {};

    // Sort edges into a cycle by following connectivity
    std::vector<HalfEdgePtr> result;

    // Find starting edge
    HalfEdgePtr current;
    if (startEdge) {
        auto it = std::find(boundaryEdges.begin(), boundaryEdges.end(), startEdge);
        if (it != boundaryEdges.end()) {
            current = startEdge;
        }
    }
    if (!current) {
        current = boundaryEdges[0];
    }

    std::set<HalfEdge*> boundarySet;
    for (const auto& e : boundaryEdges) {
        boundarySet.insert(e.get());
    }

    std::set<HalfEdge*> visited;

    while (current && visited.find(current.get()) == visited.end()) {
        visited.insert(current.get());
        result.push_back(current);

        // Find next boundary edge
        // Walk around the destination vertex to find the next boundary edge
        auto next = current->next;
        while (next && boundarySet.find(next.get()) == boundarySet.end()) {
            auto nextTwin = next->getTwin();
            if (nextTwin) {
                next = nextTwin->next;
            } else {
                break;
            }
        }

        if (next && boundarySet.find(next.get()) != boundarySet.end()) {
            current = next;
        } else {
            current = nullptr;
        }
    }

    return result;
}

std::vector<std::vector<FacePtr>> DCEL::split(const std::vector<FacePtr>& faceList) {
    if (faceList.empty()) return {};

    std::set<Face*> faceSet;
    for (const auto& f : faceList) {
        if (f) faceSet.insert(f.get());
    }

    std::set<Face*> visited;
    std::vector<std::vector<FacePtr>> components;

    for (const auto& startFace : faceList) {
        if (!startFace || visited.find(startFace.get()) != visited.end()) continue;

        // BFS to find connected component
        std::vector<FacePtr> component;
        std::queue<FacePtr> queue;
        queue.push(startFace);

        while (!queue.empty()) {
            auto current = queue.front();
            queue.pop();

            if (!current || visited.find(current.get()) != visited.end()) continue;
            visited.insert(current.get());
            component.push_back(current);

            // Add adjacent faces that are in the face set
            if (current->halfEdge) {
                auto edge = current->halfEdge;
                do {
                    auto twinEdge = edge->getTwin();
                    if (twinEdge) {
                        auto neighborFace = twinEdge->getFace();
                        if (neighborFace &&
                            faceSet.find(neighborFace.get()) != faceSet.end() &&
                            visited.find(neighborFace.get()) == visited.end()) {
                            queue.push(neighborFace);
                        }
                    }
                    edge = edge->next;
                } while (edge && edge != current->halfEdge);
            }
        }

        if (!component.empty()) {
            components.push_back(std::move(component));
        }
    }

    return components;
}

// =============================================================================
// Edge Operations
// =============================================================================

CollapseResult DCEL::collapseEdge(const HalfEdgePtr& edge) {
    CollapseResult result;
    if (!edge || !edge->origin || !edge->destination()) {
        return result;
    }

    auto v1 = edge->origin;
    auto v2 = edge->destination();

    // Compute midpoint and update v1's position (shared PointPtr mutation!)
    Point midpoint = Point::midpoint(*v1->point, *v2->point);
    v1->point->x = midpoint.x;
    v1->point->y = midpoint.y;

    result.vertex = v1;

    // Move all v2's edges to v1
    for (auto& weakEdge : v2->edges) {
        if (auto e = weakEdge.lock()) {
            e->origin = v1;
            v1->addEdge(e);
        }
    }
    v2->edges.clear();

    // Remove the collapsed edge from face cycles
    auto edgePrev = edge->getPrev();
    if (edgePrev) edgePrev->next = edge->next;
    if (edge->next) edge->next->prev = edgePrev;

    // Update face.halfEdge if it pointed to the collapsed edge
    auto edgeFace = edge->getFace();
    if (edgeFace && edgeFace->halfEdge == edge) {
        edgeFace->halfEdge = (edge->next != edge) ? edge->next : nullptr;
    }

    // Collect affected edges for face shape updates
    if (edgeFace && edge->next) {
        result.affectedEdges.push_back(edge->next);
    }

    // Handle twin edge if it exists
    auto twinEdge = edge->getTwin();
    if (twinEdge) {
        auto twinPrev = twinEdge->getPrev();
        if (twinPrev) twinPrev->next = twinEdge->next;
        if (twinEdge->next) twinEdge->next->prev = twinPrev;

        auto twinFace = twinEdge->getFace();
        if (twinFace && twinFace->halfEdge == twinEdge) {
            twinFace->halfEdge = (twinEdge->next != twinEdge) ? twinEdge->next : nullptr;
        }

        if (twinFace && twinEdge->next) {
            result.affectedEdges.push_back(twinEdge->next);
        }
    }

    // Clean up v1's edge list (remove expired references)
    v1->cleanupEdges();

    // Remove v2 from vertices map
    for (auto it = vertices.begin(); it != vertices.end(); ++it) {
        if (it->second == v2) {
            vertices.erase(it);
            break;
        }
    }

    // Remove collapsed edges from the edges vector
    edges.erase(
        std::remove_if(edges.begin(), edges.end(),
            [&edge, &twinEdge](const HalfEdgePtr& e) {
                return e == edge || e == twinEdge;
            }),
        edges.end()
    );

    return result;
}

VertexPtr DCEL::splitEdge(const HalfEdgePtr& edge) {
    if (!edge || !edge->origin || !edge->destination()) {
        return nullptr;
    }

    // Compute midpoint
    Point midpoint = Point::midpoint(*edge->origin->point, *edge->destination()->point);
    PointPtr midpointPtr = makePoint(midpoint);

    // Create new vertex
    auto newVertex = std::make_shared<Vertex>(midpointPtr);
    vertices[midpointPtr] = newVertex;

    // Create new edge from midpoint to destination
    auto newEdge = std::make_shared<HalfEdge>();

    newEdge->origin = newVertex;
    newEdge->face = edge->face;
    newEdge->next = edge->next;
    newEdge->prev = edge;

    if (edge->next) edge->next->prev = newEdge;
    edge->next = newEdge;

    newVertex->addEdge(newEdge);
    edges.push_back(newEdge);

    // Handle twin edge if it exists
    auto twinEdge = edge->getTwin();
    if (twinEdge) {
        auto newTwin = std::make_shared<HalfEdge>();

        newTwin->origin = newVertex;
        newTwin->face = twinEdge->face;
        newTwin->twin = edge;
        twinEdge->twin = newEdge;
        newEdge->twin = twinEdge;
        edge->twin = newTwin;

        newTwin->next = twinEdge->next;
        newTwin->prev = twinEdge;
        if (twinEdge->next) twinEdge->next->prev = newTwin;
        twinEdge->next = newTwin;

        newVertex->addEdge(newTwin);
        edges.push_back(newTwin);
    }

    return newVertex;
}

std::vector<HalfEdgePtr> DCEL::vertices2chain(const std::vector<VertexPtr>& verts) {
    std::vector<HalfEdgePtr> chain;
    if (verts.size() < 2) return chain;

    for (size_t i = 0; i < verts.size() - 1; ++i) {
        auto edge = findEdge(verts[i], verts[i + 1]);
        if (edge) {
            chain.push_back(edge);
        }
    }

    return chain;
}

HalfEdgePtr DCEL::findEdge(const VertexPtr& from, const VertexPtr& to) {
    if (!from || !to) return nullptr;

    for (const auto& weakEdge : from->edges) {
        if (auto edge = weakEdge.lock()) {
            if (edge->destination() == to) {
                return edge;
            }
        }
    }
    return nullptr;
}

// =============================================================================
// EdgeChain Implementation
// =============================================================================

std::vector<Point> EdgeChain::toPoly(const std::vector<HalfEdgePtr>& chain) {
    std::vector<Point> poly;
    poly.reserve(chain.size());

    for (const auto& edge : chain) {
        if (edge && edge->origin) {
            poly.push_back(*edge->origin->point);
        }
    }

    return poly;
}

std::vector<Point> EdgeChain::toPolyline(const std::vector<HalfEdgePtr>& chain) {
    std::vector<Point> poly;
    if (chain.empty()) return poly;

    poly.reserve(chain.size() + 1);

    for (const auto& edge : chain) {
        if (edge && edge->origin) {
            poly.push_back(*edge->origin->point);
        }
    }

    // Add last point (destination of final edge)
    if (!chain.empty() && chain.back()) {
        auto dest = chain.back()->destination();
        if (dest) {
            poly.push_back(*dest->point);
        }
    }

    return poly;
}

std::vector<PointPtr> EdgeChain::toPolyPtrs(const std::vector<HalfEdgePtr>& chain) {
    std::vector<PointPtr> poly;
    poly.reserve(chain.size());

    for (const auto& edge : chain) {
        if (edge && edge->origin) {
            poly.push_back(edge->origin->point);
        }
    }

    return poly;
}

std::vector<VertexPtr> EdgeChain::vertices(const std::vector<HalfEdgePtr>& chain) {
    std::vector<VertexPtr> verts;
    verts.reserve(chain.size());

    for (const auto& edge : chain) {
        if (edge) {
            verts.push_back(edge->origin);
        }
    }

    return verts;
}

HalfEdgePtr EdgeChain::edgeByOrigin(const std::vector<HalfEdgePtr>& chain, const VertexPtr& vertex) {
    for (const auto& edge : chain) {
        if (edge && edge->origin == vertex) {
            return edge;
        }
    }
    return nullptr;
}

HalfEdgePtr EdgeChain::edgeByOriginPoint(const std::vector<HalfEdgePtr>& chain, const PointPtr& point) {
    for (const auto& edge : chain) {
        if (edge && edge->origin && edge->origin->point == point) {
            return edge;
        }
    }
    return nullptr;
}

double EdgeChain::length(const std::vector<HalfEdgePtr>& chain) {
    double total = 0.0;
    for (const auto& edge : chain) {
        if (edge) {
            total += edge->length();
        }
    }
    return total;
}

bool EdgeChain::isClosed(const std::vector<HalfEdgePtr>& chain) {
    if (chain.size() < 2) return false;

    const auto& first = chain.front();
    const auto& last = chain.back();

    if (!first || !last) return false;

    return last->destination() == first->origin;
}

std::vector<HalfEdgePtr> EdgeChain::reverse(const std::vector<HalfEdgePtr>& chain) {
    std::vector<HalfEdgePtr> reversed;
    reversed.reserve(chain.size());

    for (auto it = chain.rbegin(); it != chain.rend(); ++it) {
        const auto& edge = *it;
        if (!edge) return {};

        auto twin = edge->getTwin();
        if (!twin) {
            // Can't reverse without twins
            return {};
        }
        reversed.push_back(twin);
    }

    return reversed;
}

} // namespace geom
} // namespace town_generator
